tracer/src/Datadog.Trace/Telemetry/TelemetryController.cs

```diff
@@ -0,0 +1,227 @@
+                    {
+                        Log.Debug("Process exit requested, ending telemetry loop");
+                        await PushTelemetry(isFinalPush: true).ConfigureAwait(false);
+                    }
+
+                    return;
+                }
+
+                await PushTelemetry(isFinalPush: false).ConfigureAwait(false);
+
+#if NET5_0_OR_GREATER
+                // .NET 5.0 has an explicit overload for this
+                await Task.WhenAny(
+                               Task.Delay(_sendFrequency),
+                               _tracerInitialized.Task.IsCompleted
+                                    ? _processExit.Task
+                                    : _tracerInitialized.Task)
+                          .ConfigureAwait(false);
+#else
+                tasks[1] = Task.Delay(_sendFrequency);
+                await Task.WhenAny(tasks).ConfigureAwait(false);
+#endif
+            }
+        }
+
+        private async Task PushTelemetry(bool isFinalPush)
+        {
+            try
+            {
+                var application = _configuration.GetApplicationData();
+                var host = _configuration.GetHostData();
+                if (application is null || host is null)
+                {
+                    Log.Debug("Telemetry not initialized, skipping");
+                    return;
+                }
+
+                // These calls change the state of the collectors, so must use the data
+                var success = await PushTelemetry(application, host, sendHeartbeat: !isFinalPush).ConfigureAwait(false);
+                if (!success)
+                {
+                    if (isFinalPush)
+                    {
+                        Log.Debug("Unable to send final telemetry, skipping app-closing telemetry ");
+                        return;
+                    }
+                    else
+                    {
+                        FatalError = true;
+                        Log.Debug("Unable to send telemetry, ending telemetry loop");
+                        TerminateLoop(sendAppClosingTelemetry: false);
+                    }
+                }
+
+                if (isFinalPush)
+                {
+                    var closingTelemetryData = _dataBuilder.BuildAppClosingTelemetryData(application, host);
+
+                    Log.Debug("Pushing app-closing telemetry");
+                    await _transport.PushTelemetry(closingTelemetryData).ConfigureAwait(false);
+                }
+            }
+            catch (Exception ex)
+            {
+                Log.Warning(ex, "Error pushing telemetry");
+            }
+        }
+
+        private async Task<bool> PushTelemetry(ApplicationTelemetryData application, HostTelemetryData host, bool sendHeartbeat)
+        {
+            // use values from previous failed attempt if necessary
+            var configuration = _configuration.GetConfigurationData() ?? _circuitBreaker.PreviousConfiguration;
```

Comment:
It looks like we only send a previously failed request if the current configuration is null (aka no changes). Is that the intent or should there be another mechanism to retry previously failed sequences?

Reply:
Yeah, that's correct. `_configuration.GetConfigurationData()` returns the _latest_ settings, or null if there's no changes. We always want to send the latest data (the backend essentially overwrites previous versions anyway, so there's no benefit in sending both)